zebra-consensus/src/block/check.rs

```diff
@@ -53,6 +57,50 @@ pub fn is_equihash_solution_valid(header: &Header) -> Result<(), equihash::Error
 /// accepted." [§7.5][7.5]
 ///
 /// [7.5]: https://zips.z.cash/protocol/protocol.pdf#blockheader
-pub fn is_time_valid_at(header: &Header, now: DateTime<Utc>) -> Result<(), BoxError> {
+pub fn time_is_valid_at(header: &Header, now: DateTime<Utc>) -> Result<(), BoxError> {
     header.is_time_valid_at(now)
 }
+
+/// [3.9]: https://zips.z.cash/protocol/protocol.pdf#subsidyconcepts
+pub fn subsidy_is_correct(network: Network, block: &Block) -> Result<(), BlockError> {
+    let height = block
+        .coinbase_height()
+        .expect("always called on blocks with a coinbase height");
+    let halving_div = subsidy::general::halving_divisor(height, network);
+
+    let coinbase = block.transactions.get(0).ok_or(SubsidyError::NoCoinbase)?;
+
+    // TODO: the sum of the coinbase transaction outputs must be less than or equal to the block subsidy plus transaction fees
+
+    // Check founders reward and funding streams
+    match (halving_div, height) {
+        (_, height) if (height < SLOW_START_INTERVAL) => unreachable!(
+            "unsupported block height: callers should handle blocks below {:?}",
+            SLOW_START_INTERVAL
+        ),
+
+        (halving_div, _) if (halving_div.count_ones() != 1) => unreachable!(
+            "invalid halving divisor: the halving divisor must be a non-zero power of two"
+        ),
+
+        (1, _) => {
+            // validate founders reward
+            let founders_reward = subsidy::founders_reward::founders_reward(height, network)
+                .expect("invalid Amount: founders reward should be valid");
+            let matching_values =
+                subsidy::general::find_output_with_amount(coinbase, founders_reward);
+
+            // TODO: the exact founders reward value must be sent as a single output to the correct address
+            if !matching_values.is_empty() {
+                Ok(())
+            } else {
+                Err(SubsidyError::FoundersRewardNotFound)?
+            }
+        }
+
+        (2, _) => unimplemented!("funding stream block subsidy validation is not implemented"),
+
+        // Valid halving, with no founders reward or funding streams
+        _ => Ok(()),
+    }
```

Comment:
I think it would be more appropriate to write this as a series of if statements.

Reply:
This is an initial implementation, in future PRs, we'll be adding cases for:
* shielded coinbase
* funding streams

I think the need for a match statement will be much clearer after those changes - but I can put this change on the final cleanup list, if you'd like?

Reply:
I'm not really familiar with shielded coinbase or funding streams and how they'd impact this. How do those changes make the match statement the clearer option?

Reply:
Shielded coinbase activated with Heartwood, so it splits the first halving into two separate cases (or modifies the code in that block).

Funding streams will fill in the code for the second halving.

Maybe an if statement would work here, if we refactored the larger blocks into functions. Then the if statement itself would be more readable.

I'm also used to using match with Height, because we couldn't do Height comparisons until this PR.

I'll see how I go with this.

Reply:
Done in https://github.com/oxarbitrage/zebra/pull/30

Added a cleanup task to #801: Split `subsidy_is_correct` into non-miner, founders, and funding stream validation, if needed